### PR TITLE
improve: Virtualize mo.ui.multiselect to handle large lists

### PR DESCRIPTION
**Pull request by wasimsandhu**
## 📝 Summary

Completes #588.

Screenshot 2024-06-20 at 12 42 22 AM

## 🔍 Description of Changes

Installed `[react-virtuoso](https://github.com/petyosi/react-virtuoso)` and passed `<Virtuoso />` to `<Combobox />`.

Because `cmdk` does not support filtering with virtualization, I used the existing filter function in `MultiselectPlugin.tsx` to filter the options array directly.

## 📋 Checklist

- [x] I have read the [contributor guidelines](../CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://discord.gg/JE7nhX6mD8), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [x] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

@mscolnick

### CODE REVIEW

**Comment by wasimsandhu** [comment]
After wrestling with `@tanstack/react-virtual` for a few hours, using Virtuoso was surprisingly easy. 

Installed [`react-virtuoso`](https://github.com/petyosi/react-virtuoso) and passed `<Virtuoso />` to `<Combobox />`.

It broke `cmdk` filtering since the `<ComboboxItem>` components are wrapped in the virtuoso element:

```
TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.
```

The `cmdk` [FAQ](https://github.com/pacocoursey/cmdk?tab=readme-ov-file#faq) recommends implementing filtering on your own for virtualization anyway, so that's what I did.

**Comment by wasimsandhu** [comment]
Tested multiselect on my own machine and it's buttery smooth now -- did not spot any regressions.

I did see that `<Combobox>` is used in `switchable-multi-select` and `form.tsx` -- I did not get a chance to do regression tests for those.

File: frontend/src/components/ui/combobox.tsx [hunk 1]
```diff
@@ -187,7 +190,7 @@ export const Combobox = <TValue,>({
               rootClassName={"px-1 h-8"}
               autoFocus={true}
               value={search}
-              onValueChange={onSearchChange}
+              onValueChange={customFilterFn ?? onSearchChange}
```
Comment (mscolnick) [inline frontend/src/components/ui/combobox.tsx:193]: we can probably re-use the `onSearchChange` above - and remove the need for `customFilterFn`

File: frontend/src/plugins/impl/MultiselectPlugin.tsx [hunk 2]
```diff
@@ -53,6 +54,24 @@ interface MultiselectProps extends Data {
...
+    if (!search) {
+      setFilteredOptions(props.options);
+      return;
+    }
+    setFilteredOptions(
+      props.options.filter(
+        (option) => multiselectFilterFn(option, search) === 1,
+      ),
+    );
+  };
```
Comment (mscolnick) [inline frontend/src/plugins/impl/MultiselectPlugin.tsx:74]: the react team suggests minimizing the use of useEffect, you could re-write this as:

```ts
  const [searchQuery, setSearchQuery] = useState<string>("");
  const filteredResults = useMemo(() => {
      if (!searchQuery) {
         return props.options;
      }
      return props.options.filter(
          (option) => multiselectFilterFn(option, search) === 1,
      );
  }, [props.options, searchQuery]);
```

File: frontend/src/plugins/impl/MultiselectPlugin.tsx [hunk 3]
```diff
@@ -67,18 +86,25 @@ const Multiselect = (props: MultiselectProps): JSX.Element => {
...
         placeholder="Select..."
         multiple={true}
-        filterFn={multiselectFilterFn}
         className={cn({
           "w-full": props.fullWidth,
         })}
         value={props.value}
         onValueChange={(newValues) => setValue(newValues || [])}
+        shouldFilter={false}
+        customFilterFn={filterOptions}
```
Comment (mscolnick) [inline frontend/src/plugins/impl/MultiselectPlugin.tsx:95]: can change this to `onSearchChange={setSearchQuery}`

File: frontend/src/plugins/impl/MultiselectPlugin.tsx [hunk 4]
```diff
@@ -67,18 +86,25 @@ const Multiselect = (props: MultiselectProps): JSX.Element => {
...
+        <Virtuoso
+          style={{ height: "200px" }}
+          totalCount={filteredOptions.length}
+          overscan={25}
+          itemContent={(i: number) => (
+            <ComboboxItem key={filteredOptions[i]} value={filteredOptions[i]}>
+              {filteredOptions[i]}
+            </ComboboxItem>
+          )}
+        />
```
Comment (mscolnick) [inline frontend/src/plugins/impl/MultiselectPlugin.tsx:107]: did you notice any downside if the size is 10 (and we don't need to virtualize?)
if there is no downside, we can always virtualize

but if it feels worse for small numbers, we do do something like:

```ts
const renderList = () {

  if (options.length > 100) {
    return <Virtuoso
  }
  return  props.options.map((option) => (
          <ComboboxItem key={option} value={option}>
            {option}
          </ComboboxItem>
        ))}
}
```

File: marimo/_plugins/ui/_impl/input.py [hunk 5]
```diff
@@ -914,17 +914,6 @@ def __init__(
         full_width: bool = False,
         max_selections: Optional[int] = None,
     ) -> None:
-        if len(options) > multiselect._MAX_OPTIONS:
```
Comment (mscolnick) [inline marimo/_plugins/ui/_impl/input.py:917]: can you try a _really_ large list (e.g. 10,000)? maybe we still need this, but can just bump it higher. 

I saw you commited a notebook - https://github.com/marimo-team/marimo/pull/1654/commits/5471d9f0b96f9cf1f8fd975bae48bdf7c3d96ac5
we actually like doing this - and putting them in `mairmo/_smoke_tests/bugs/<issue_number_or_name>.py`

File: frontend/src/plugins/impl/MultiselectPlugin.tsx (hunk 2 above)
Comment (wasimsandhu) [reply frontend/src/plugins/impl/MultiselectPlugin.tsx:74]: TIL, thanks for letting me know

File: marimo/_plugins/ui/_impl/input.py (hunk 5 above)
Comment (wasimsandhu) [reply marimo/_plugins/ui/_impl/input.py:917]: Oh cool! Yeah let me test with small / large lists and add it back then.

File: frontend/src/plugins/impl/MultiselectPlugin.tsx (hunk 4 above)
Comment (wasimsandhu) [reply frontend/src/plugins/impl/MultiselectPlugin.tsx:107]: Just checked with 50 items and it feels like the list is sort of jumping when I scroll. I think this is a good idea.

Comment (mscolnick) [reply frontend/src/plugins/impl/MultiselectPlugin.tsx:107]: cool - ill defer to you on the number. maybe 200?

Comment (wasimsandhu) [reply frontend/src/plugins/impl/MultiselectPlugin.tsx:107]: I think 200 is good. To some extent, there is some stuttering / flickering with 10,000 items if you do a big scroll and let the list slow to a stop. But I don't think it's too noticeable with controlled scrolls, and better than a laggy list / bloated DOM anyway.

Comment (wasimsandhu) [reply frontend/src/plugins/impl/MultiselectPlugin.tsx:107]: I'll commit my test notebook in a sec and let you be the judge

File: marimo/_plugins/ui/_impl/input.py (hunk 5 above)
Comment (wasimsandhu) [reply marimo/_plugins/ui/_impl/input.py:917]: Seems fine! Tested with 10,000 items and 100,000 items. However, with 1M items:
```
Your output is too large for marimo to show. It has a size of 23889316 bytes. Did you output this object by accident?
```
I'll add this limit and its test back in.

File: frontend/src/components/ui/combobox.tsx [hunk 6]
```diff
@@ -45,7 +45,7 @@ type ComboboxFilterProps =
       filterFn?: React.ComponentProps<typeof Command>["filter"];
     }
   | {
-      shouldFilter: false;
+      shouldFilter?: false;
```
Comment (mscolnick) [inline frontend/src/components/ui/combobox.tsx:48]: revert?

File: marimo/_smoke_tests/bugs/1654-virtualize-multiselect.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+# Copyright 2024 Marimo. All rights reserved.
```
Comment (mscolnick) [inline marimo/_smoke_tests/bugs/1654-virtualize-multiselect.py:1]: nice test!

File: frontend/src/components/ui/combobox.tsx (hunk 6 above)
Comment (wasimsandhu) [reply frontend/src/components/ui/combobox.tsx:48]: whoops missed it

**Review by mscolnick — APPROVED** [review]

**Comment by mscolnick** [comment]
awesome work @wasimsandhu!

**Pull request merged**